src/lib/cached/cached.ml

```diff
@@ -126,68 +135,150 @@ module Spec = struct
         ; autogen_path: string
         ; manual_install_path: string
         ; brew_install_path: string
+        ; s3_install_path: string
         ; digest_input: 'input -> string
         ; create_env: 'input -> 'env
         ; input: 'input }
         -> 'a t
 
   let create ~load ~name ~autogen_path ~manual_install_path ~brew_install_path
-      ~digest_input ~create_env ~input =
+      ~s3_install_path ~digest_input ~create_env ~input =
     T
       { load
       ; name
       ; autogen_path
       ; manual_install_path
       ; brew_install_path
+      ; s3_install_path
       ; digest_input
       ; create_env
       ; input }
 end
 
+module Track_generated = struct
+  type t = [`Generated_something | `Cache_hit]
+
+  let ( + ) x y =
+    match (x, y) with
+    | `Generated_something, _ ->
+        `Generated_something
+    | _, `Generated_something ->
+        `Generated_something
+    | `Cache_hit, `Cache_hit ->
+        `Cache_hit
+end
+
 let run
     (Spec.T
       { load
       ; name
       ; autogen_path
       ; manual_install_path
       ; brew_install_path
+      ; s3_install_path
       ; digest_input
       ; create_env
       ; input }) =
   let open Deferred.Let_syntax in
   let hash = digest_input input in
+  let s3_bucket_prefix =
+    "https://s3-us-west-2.amazonaws.com/snark-keys.o1test.net"
+  in
   let base_path directory = directory ^/ hash in
+  let full_paths directory =
+    With_components.path load ~base_path:(base_path directory)
+  in
   match%bind
-    Deferred.List.fold [manual_install_path; brew_install_path] ~init:None
+    Deferred.List.fold
+      [manual_install_path; brew_install_path; s3_install_path] ~init:None
       ~f:(fun acc path ->
         if is_some acc then return acc
         else
           match%map With_components.load load ~base_path:(base_path path) with
           | Ok x ->
-              Core.printf "Loaded %s from the following path %s\n" name path ;
+              Core.printf
```

Comment:
is it better to use `Core_kernel` than `Core` for these printfs (smaller dependency)?